File: workvpn-ios/WorkVPN/Utils/OVPNParser.swift
//
//  OVPNParser.swift
//  WorkVPN
//
//  Parser for .ovpn configuration files
//

import Foundation

enum OVPNParserError: LocalizedError {
    case missingRemote
    case missingCA
    case invalidFormat
    case invalidPort

    var errorDescription: String? {
        switch self {
        case .missingRemote:
            return "Configuration is missing remote server address"
        case .missingCA:
            return "Configuration is missing CA certificate"
        case .invalidFormat:
            return "Invalid OpenVPN configuration format"
        case .invalidPort:
            return "Invalid port number"
        }
    }
}

struct OVPNParser {
    static func parse(content: String, name: String) throws -> VPNConfig {
        let lines = content.components(separatedBy: .newlines)

        var serverAddress: String?
        var port: Int = 1194
        var `protocol`: String = "udp"
        var ca: String?
        var cert: String?
        var key: String?
        var tlsAuth: String?
        var cipher: String?
        var auth: String?

        var currentBlock: String?
        var blockContent: [String] = []

        for line in lines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)

            // Skip empty lines and comments
            if trimmed.isEmpty || trimmed.hasPrefix("#") || trimmed.hasPrefix(";") {
                continue
            }

            // Handle inline blocks
            if trimmed.hasPrefix("<") {
                if let match = trimmed.range(of: #"<(\/?)([

\w-]+)>"#, options: .regularExpression) {
                    let tag = String(trimmed[match])
                    let isClosing = tag.hasPrefix("</")
                    let blockType = tag.replacingOccurrences(of: "<", with: "")
                        .replacingOccurrences(of: "</", with: "")
                        .replacingOccurrences(of: ">", with: "")

                    if isClosing {
                        // End of block
                        if currentBlock == blockType {
                            let content = blockContent.joined(separator: "\n")

                            switch blockType {
                            case "ca":
                                ca = content
                            case "cert":
                                cert = content
                            case "key":
                                key = content
                            case "tls-auth":
                                tlsAuth = content
                            default:
                                break
                            }

                            currentBlock = nil
                            blockContent = []
                        }
                    } else {
                        // Start of block
                        currentBlock = blockType
                        blockContent = []
                    }
                }
                continue
            }

            // If in block, collect content
            if currentBlock != nil {
                blockContent.append(line)
                continue
            }

            // Parse key-value pairs
            let parts = trimmed.components(separatedBy: .whitespaces)
            guard let key = parts.first else { continue }
            let values = Array(parts.dropFirst())

            switch key {
            case "remote":
                if values.count > 0 {
                    serverAddress = values[0]
                    if values.count > 1, let portValue = Int(values[1]) {
                        port = portValue
                    }
                }

            case "proto":
                if let proto = values.first {
                    `protocol` = proto
                }

            case "cipher":
                if let cipherValue = values.first {
                    cipher = cipherValue
                }

            case "auth":
                if let authValue = values.first {
                    auth = authValue
                }

            default:
                break
            }
        }

        // Validation
        guard let server = serverAddress else {
            throw OVPNParserError.missingRemote
        }

        if ca == nil {
            throw OVPNParserError.missingCA
        }

        // Create config
        var config = VPNConfig(
            name: name,
            content: content,
            serverAddress: server,
            port: port,
            protocol: `protocol`
        )

        config.ca = ca
        config.cert = cert
        config.key = key
        config.tlsAuth = tlsAuth
        config.cipher = cipher
        config.auth = auth

        return config
    }

    static func validate(config: VPNConfig) -> [String] {
        var errors: [String] = []

        if config.serverAddress.isEmpty {
            errors.append("Missing remote server address")
        }

        if config.ca == nil {
            errors.append("Missing CA certificate")
        }

        if config.port < 1 || config.port > 65535 {
            errors.append("Invalid port number")
        }

        return errors
    }
}

File: workvpn-android/app/src/main/java/com/workvpn/android/util/OVPNParser.kt
package com.workvpn.android.util

import com.workvpn.android.model.VPNConfig

object OVPNParser {

    sealed class ParseError : Exception() {
        object MissingRemote : ParseError()
        object MissingCA : ParseError()
        object InvalidFormat : ParseError()
        object InvalidPort : ParseError()

        override val message: String
            get() = when (this) {
                is MissingRemote -> "Configuration is missing remote server address"
                is MissingCA -> "Configuration is missing CA certificate"
                is InvalidFormat -> "Invalid OpenVPN configuration format"
                is InvalidPort -> "Invalid port number"
            }
    }

    fun parse(content: String, name: String): VPNConfig {
        val lines = content.lines()

        var serverAddress: String? = null
        var port: Int = 1194
        var protocol: String = "udp"
        var deviceType: String = "tun"
        var ca: String? = null
        var cert: String? = null
        var key: String? = null
        var tlsAuth: String? = null
        var cipher: String? = null
        var auth: String? = null

        var currentBlock: String? = null
        val blockContent = mutableListOf<String>()

        for (line in lines) {
            val trimmed = line.trim()

            // Skip empty lines and comments
            if (trimmed.isEmpty() || trimmed.startsWith("#") || trimmed.startsWith(";")) {
                continue
            }

            // Handle inline blocks
            if (trimmed.startsWith("<")) {
                val blockPattern = Regex("""<(/?)([\\w-]+)>""")
                val match = blockPattern.find(trimmed)

                if (match != null) {
                    val isClosing = match.groupValues[1] == "/"
                    val blockType = match.groupValues[2]

                    if (isClosing) {
                        // End of block
                        if (currentBlock == blockType) {
                            val content = blockContent.joinToString("\n")

                            when (blockType) {
                                "ca" -> ca = content
                                "cert" -> cert = content
                                "key" -> key = content
                                "tls-auth" -> tlsAuth = content
                            }

                            currentBlock = null
                            blockContent.clear()
                        }
                    } else {
                        // Start of block
                        currentBlock = blockType
                        blockContent.clear()
                    }
                }
                continue
            }

            // If in block, collect content
            if (currentBlock != null) {
                blockContent.add(line)
                continue
            }

            // Parse key-value pairs
            val parts = trimmed.split(Regex("\\s+"))
            val keyword = parts.firstOrNull() ?: continue
            val values = parts.drop(1)

            when (keyword) {
                "remote" -> {
                    if (values.isNotEmpty()) {
                        serverAddress = values[0]
                        if (values.size > 1) {
                            port = values[1].toIntOrNull() ?: 1194
                        }
                    }
                }
                "proto" -> {
                    if (values.isNotEmpty()) {
                        protocol = values[0]
                    }
                }
                "dev" -> {
                    if (values.isNotEmpty()) {
                        deviceType = values[0]
                    }
                }
                "cipher" -> {
                    if (values.isNotEmpty()) {
                        cipher = values[0]
                    }
                }
                "auth" -> {
                    if (values.isNotEmpty()) {
                        auth = values[0]
                    }
                }
            }
        }

        // Validation
        if (serverAddress == null) {
            throw ParseError.MissingRemote
        }

        if (ca == null) {
            throw ParseError.MissingCA
        }

        return VPNConfig(
            name = name,
            content = content,
            serverAddress = serverAddress,
            port = port,
            protocol = protocol,
            ca = ca,
            cert = cert,
            key = key,
            tlsAuth = tlsAuth,
            cipher = cipher,
            auth = auth,
            deviceType = deviceType
        )
    }

    fun validate(config: VPNConfig): List<String> {
        val errors = mutableListOf<String>()

        if (config.serverAddress.isBlank()) {
            errors.add("Missing remote server address")
        }

        if (config.ca == null) {
            errors.add("Missing CA certificate")
        }

        if (config.port !in 1..65535) {
            errors.add("Invalid port number")
        }

        return errors
    }
}
